pac4j-oidc/src/main/java/org/pac4j/oidc/credentials/authenticator/OidcAuthenticator.java

```diff
@@ -50,17 +58,40 @@ public OidcAuthenticator(final OidcConfiguration configuration, final OidcClient
 
         final ClientAuthenticationMethod chosenMethod;
         final ClientAuthenticationMethod configurationMethod = configuration.getClientAuthenticationMethod();
+        
+        final ClientAuthenticationMethod preferredMethod;
+        if (configurationMethod == null || SUPPORTED_METHODS.contains(configurationMethod)) {
+            preferredMethod = configurationMethod;
+        } else {
+            // Use method suggested by provider metadata.
+            preferredMethod = null;
+            logger.warn("Configured authentication method ({}) is not supported.", configurationMethod);
+        } 
+        
         if (CommonHelper.isNotEmpty(metadataMethods)) {
-            if (metadataMethods.contains(configurationMethod)) {
-                chosenMethod = configurationMethod;
+            if (preferredMethod != null && metadataMethods.contains(preferredMethod)) {
+                chosenMethod = preferredMethod;
             } else {
-                chosenMethod = metadataMethods.get(0);
```

Comment:
I found your code a bit complicated: isn't it this line only we should change to take the first supported authentication methods?

Reply:
There are some more problems: 
* When only changing this line, I still get a warning, if I do not configure a client authentication method, see the code following this line. This is not acceptable default behavior.
* In the case `!CommonHelper.isNotEmpty(metadataMethods)` (if the provider metadata does not provide any information of provider-supported methods), the code did choose `ClientAuthenticationMethod.getDefault()`, even if I have explicitly configured a method to use. This is quite absurd, because I'm not able to explicitly choose a method in case of missing metadata information from a provider. 
* The problem, that the explicitly configured method is not supported by the implementation should be handled separately resulting in an explicity warning message.
* The fact that the configured method could be `null` should be made explicit. In my opinion, a contains-check `metadataMethods.contains(configurationMethod)` should not be done with the argument being `null`, where `null` has a special meaning "I want to select the best matching provider-supported method".

Reply:
You're right. But I'd still prefer to simplify the source code with the explicit four use cases:metadata yes/no, configurationMethod: null/not null.

```java
if (metadata) {
  if (configMethod != null) {
    if (! metadata.contains(configMethod)) throw Exception
  } else {
    configMethod = metadata.search(client_secret_basic or client_secret_post)
    if (configMethod == null) throw Exception
  }
} else (no metadata) {
  if (configMethod != null) {
    if (configMethod != client_secret_basic and configMethod != client_secret_post) throw Exception
  } else {
    configMethod = client_secret_basic
  }
}
```

Reply:
In case 
```
if (metadata) {
  if (configMethod != null) {
    if (metadata.contains(configMethod)) {
```
The configured method may still not within the supported methods of the pac4j library. This is the reason, why I separated the lookup of the preferred method from the rest of the logic.

I'll try to re-formulate with your suggested test-sequence.